Sources/SwiftFusion/Core/Manifold.swift

```diff
@@ -0,0 +1,110 @@
+public protocol ManifoldCoordinate: Differentiable {
+  /// The local coordinate system used in the chart.
+  ///
+  /// Isomorphic to `R^n`, where `n` is the dimension of the manifold.
+  associatedtype LocalCoordinate: AdditiveArithmetic & Differentiable & VectorProtocol
+    where LocalCoordinate.TangentVector == LocalCoordinate
+
+  /// The global coordinate corresponding to `local` in the chart centered around `self`.
+  ///
+  /// Satisfies the following properties:
+  /// - `global(LocalCoordinate.zero) == self`
+  /// - There exists an open set `B` around `LocalCoordinate.zero` such that
+  ///   `local(global(b)) == b` for all `b \in B`.
+  @differentiable(wrt: local)
+  func global(_ local: LocalCoordinate) -> Self
+
+  /// The local coordinate corresponding to `global` in the chart centered around `self`.
+  ///
+  /// Satisfies the following properties:
+  /// - `local(self) == LocalCoordinate.zero`
+  /// - There exists an open set `B` around `self` such that `local(global(b)) == b` for all
+  ///   `b \in B`.
+  @differentiable(wrt: global)
+  func local(_ global: Self) -> LocalCoordinate
+}
+
+/// A point on a differentiable manifold.
+public protocol Manifold: Differentiable {
+  /// The manifold's global coordinate system.
+  associatedtype Coordinate: ManifoldCoordinate
+      where Coordinate.LocalCoordinate == Self.TangentVector
+
+  /// The coordinate of `self`.
+  ///
+  /// Note: This is not differentiable, and therefore clients should use `coordinate` instead.
+  var coordinateStorage: Coordinate { get set }
+
+  /// Creates a manifold point with coordinate `coordinateStorage`.
+  ///
+  /// Note: This is not differentiable, and therefore clients should use `init(coordinate:)`
+  /// instead.
+  init(coordinateStorage: Coordinate)
+}
+
+extension Manifold {
+  /// The coordinate of `self`.
+  @differentiable
+  public var coordinate: Coordinate { coordinateStorage }
+
+  /// A custom derivative of `coordinate` that converts from the global coordinate system's
+  /// tangent vector to the local coordinate system's tangent vector, so that all functions on this
+  /// manifold using `coordinate` have derivatives involving local coordinates.
+  @derivative(of: coordinate)
+  @usableFromInline
+  func vjpCoordinate() -> (value: Coordinate, pullback: (Coordinate.TangentVector) -> TangentVector) {
+
+    // Explanation of this pullback:
+    //
+    // Let `f: Manifold -> Coordinate` be `f(x) = x.coordinateStorage`.
+    //
+    // `D_x(f)` (the derivative of `f` at `x`) is a linear approximation of how changes in local
+    // coordinates around `x` lead to changes in global coordinates around `x.coordinateStorage`.
+    //
+    // `x.coordinateStorage.global: LocalCoordinate -> Coordinate` defines _exactly_ how local
+    // coordinates at `x` map to global coordinates.
+    //
+    // Therefore, `D_x(f)` is the derivative of `x.coordinateStorage.global`.
```

Comment:
Yes, but close the loop :-) Why does that mean that `coordinateStorage.global` is the pullback? And (see confusion above) why is it not `coordinate.global` ?

Reply:
I have an idea how to update this whole comment to be clearer by using notation closer to the Swift code and by explicitly saying how we get the pullback expression:

```
    // Explanation of this pullback:
    //
    // Let `f: Manifold -> Coordinate` be `f(x) = x.coordinateStorage`.
    //
    // `differential(at: x, in: f)` is a linear approximation of how changes in local
    // coordinates around `x` lead to changes in global coordinates around `x.coordinateStorage`.
    //
    // `x.coordinateStorage.global: LocalCoordinate -> Coordinate` defines _exactly_ how local
    // coordinates around zero map to global coordinates around `x`.
    //
    // Therefore, `differential(at: x, in: f) = differential(at: zero, in: x.coordinateStorage.global)`.
    //
    // The pullback is the dual map of the differential, so taking duals of both sides gives:
    //   `pullback(at: x, in: f) = pullback(at: zero, in: x.coordinateStorage.global)`.
```